proposals/lip-0053.md

```diff
@@ -133,28 +144,247 @@ The following constants are used throughout the document:
+        for currentValidator in activeValidators:
+            if currentValidator.blsKey == validator.blsKey:
+                currentValidator.bftWeight = validator.bftWeight
+        else:
+            add validator to activeValidators maintaining the entries in lexicographical order of blsKeys
+
+    #remove any entry activeValidators with bftWeight == 0
+    for validator in activeValidators:
+        if validator.bftWeight == 0:
+            activeValidators.remove(validator)
+    return activeValidators
+```
+
+#### verifyChannel
+
+```python
+def verifyChannel(trs: Transaction) -> None:
+    partnerchainID = trs.params.sendingChainID
+    inboxTree = Merkle tree of channel(partnerchainID).inbox
+    # Update inbox by appending all cross-chain messages.
+    # crossChainMessages can be empty.
+    for ccm in trs.params.inboxUpdate.crossChainMessages:
+        inboxTree.append(SHA-256(appendData))
+    newInboxRoot = inboxTree.root
+
+    # If the message witness is non-empty, update the inbox root.
+    messageWitnessHashes = trs.params.inboxUpdate.messageWitnessHashes
+    if len(messageWitnessHashes) > 0:
+        # calculateRootFromRightWitness is specified in LIP 0031.
+        newInboxRoot = calculateRootFromRightWitness(inboxTree.size, inboxTree.appendPath, messageWitnessHashes)
+
+    # Verification for non-empty certificates.
+    if len(trs.params.certificate) > 0:
+        outboxRootWitness = trs.params.inboxUpdate.outboxRootWitness
+        outboxKey = MODULE_ID_INTEROPERABILITY + STORE_PREFIX_OUTBOX_ROOT + SHA-256(partnerchainID)
+        proof = {
+            siblingHashes: outboxRootWitness.siblingHashes,
+            queries: [{
+                key: outboxKey,
+                value: newInboxRoot,
+                bitmap: outboxRootWitness.bitmap
+            }],
+        }
+        
+        # verify is specified in LIP 0039.
+        certificate = decode(certificateSchema, trs.params.certificate)
+        if verify([outboxKey], proof, certificate.stateRoot) == False:
+            raise Exception("Invalid inclusion proof for inbox update.")
+    
+    # Verification for empty certificates.
+    else:
+        if newInboxRoot != channel(partnerchainID).partnerChainOutboxRoot:
+            raise Exception("Inbox root does not match partner chain outbox root.")
+```
+
+#### updateValidators
+
+```python
+def updateValidators(trs: Transaction) -> None:
+    if trs.params.newCertificateThreshold > 0:
+        validators(trs.params.sendingChainID).certificateThreshold = trs.params.newCertificateThreshold
+
+    validators(trs.params.sendingChainID).activeValidators = updateActiveValidators(
+        validators(trs.params.sendingChainID).activeValidators, 
+        trs.params.activeValidatorsUpdate
+    )
+```
+
+#### updateCertificate
+
+```python
+def updateCertificate(trs: Transaction) -> None:
+    certificate = decode(certificateSchema, trs.params.certificate)
+    sidechainAccount.lastCertificate = {
```

Comment:
Should we modify the chain data substore here?

Reply:
I had a typo in that sidechainAccount should be chainAccount(ccu.params.sendingChainID)

Reply:
Ah, now it is clear.